### PR TITLE
Fix deadlock between map lock and snapshot lock

### PR DESCRIPTION
**Pull request by takuro-sato**
## Cause of bug

https://github.com/microsoft/CCF/issues/5051

# Fix

Use consistent order of locks. First MAP_LOCK then SNAPSHOT_LOCK.
I didn't choose the opposite because the fix would be much more complicated.

# Test

- Run reconfiguration test which caused the bug in https://github.com/microsoft/CCF/issues/5051 more than 20 times.
- Enable `TSAN` option and comment out deadlock suppression in `tsan_env_suppressions`, and then run tests.sh. Confirmed there is no warning about order between MAP_LOCK andSNAPSHOT_LOCK.
[test_log.zip](https://github.com/microsoft/CCF/files/10899483/test_log.zip)

### CODE REVIEW

File: cmake/common.cmake [hunk 1]
```diff
@@ -93,6 +94,13 @@ if(SAN AND LVI_MITIGATIONS)
   )
 endif()
 
+if(TSAN AND LVI_MITIGATIONS)
+  message(
+    FATAL_ERROR
+      "Building with both TSAN and LVI mitigations is unsafe and deadlocks - choose one"
```
Comment (achamayou) [inline cmake/common.cmake:100]: It doesn't deadlock, but causes the linker to get stuck in an infinite loop.

Comment (takuro-sato) [reply cmake/common.cmake:100]: Thanks for clarifying. This PR is based on https://github.com/microsoft/CCF/pull/5073 and this change is a part of it.
I wanted to use TSAN to check if the changes are OK.

File: src/node/snapshotter.h [hunk 2]
```diff
@@ -357,6 +359,8 @@ namespace ccf
       // at the last snapshottable index before idx, and schedule snapshot
       // serialisation on another thread (round-robin). Otherwise, only record
       // that a snapshot was generated.
+
+      kv::ScopedStoreMapsLock maps_lock(store);
```
Comment (jumaffre) [inline src/node/snapshotter.h:363]: Do we know the place in the code these two locks are acquired in the inverse order, leading to a deadlock?

Comment (takuro-sato) [reply src/node/snapshotter.h:363]: After the fix, I believe there is no more inverse order between map lock and snapshot lock based on TSAN result.

**Review by jumaffre — APPROVED** [review]

File: src/node/snapshotter.h (hunk 2 above)
Comment (achamayou) [reply src/node/snapshotter.h:363]: I'd be curious too.

Comment (takuro-sato) [reply src/node/snapshotter.h:363]: Please check https://github.com/microsoft/CCF/issues/5051

**Pull request merged**